Sources/NIOCore/NIOScheduledCallback.swift

```diff
@@ -0,0 +1,166 @@
+///
+/// - Seealso: `EventLoop.scheduleCallback(at:handler:)`.
+public protocol NIOScheduledCallbackHandler {
+    /// This function is called at the scheduled time, unless the scheduled callback is cancelled.
+    ///
+    /// - Parameter eventLoop: The event loop on which the callback was scheduled.
+    func handleScheduledCallback(eventLoop: some EventLoop)
+
+    /// This function is called if the scheduled callback is cancelled.
+    ///
+    /// The callback could be cancelled explictily, by the user calling ``NIOScheduledCallback/cancel()``, or
+    /// implicitly, if it was still pending when the event loop was shut down.
+    ///
+    /// - Parameter eventLoop: The event loop on which the callback was scheduled.
+    func onCancelScheduledCallback(eventLoop: some EventLoop)
+}
+
+extension NIOScheduledCallbackHandler {
+    /// Default implementation of `onCancelScheduledCallback(eventLoop:)`: does nothing.
+    public func onCancelScheduledCallback(eventLoop: some EventLoop) {}
+}
+
+/// An opaque handle that can be used to cancel a scheduled callback.
+///
+/// Users should not create an instance of this type; it is returned by `EventLoop.scheduleCallback(at:handler:)`.
+///
+/// - Seealso: `EventLoop.scheduleCallback(at:handler:)`.
+public struct NIOScheduledCallback: Sendable {
+    @usableFromInline
+    enum Backing: Sendable {
+        /// A task created using `EventLoop.scheduleTask(deadline:_:)` by the default implementation.
+        case `default`(_ task: Scheduled<Void>)
+        /// A custom callback identifier, used by event loops that provide a custom implementation.
+        case custom(id: UInt64)
+    }
+
+    @usableFromInline
+    var eventLoop: any EventLoop
+
+    @usableFromInline
+    var backing: Backing
+
+    /// This initializer is only for the default implementation and is fileprivate to avoid use in EL implementations.
+    fileprivate init(_ eventLoop: any EventLoop, _ task: Scheduled<Void>) {
+        self.eventLoop = eventLoop
+        self.backing = .default(task)
+    }
+
+    /// Create a handle for the scheduled callback with an opaque identifier managed by the event loop.
+    ///
+    /// - NOTE: This initializer is for event loop implementors only, end users should use `EventLoop.scheduleCallback`.
+    ///
+    /// - Seealso: `EventLoop.scheduleCallback(at:handler:)`.
+    @inlinable
+    public init(_ eventLoop: any EventLoop, id: UInt64) {
+        self.eventLoop = eventLoop
+        self.backing = .custom(id: id)
+    }
+
+    /// Cancel the scheduled callback associated with this handle.
+    @inlinable
+    public func cancel() {
+        self.eventLoop.cancelScheduledCallback(self)
+    }
+
+    /// The callback identifier, if the event loop uses a custom scheduled callback implementation; nil otherwise.
+    ///
+    /// - NOTE: This property is for event loop implementors only.
+    @inlinable
+    public var customCallbackID: UInt64? {
```

Comment:
NIT: Do we need the `custom` here in the naming?

Reply:
IMO it adds value when glancing at it as the property name implies that it's only relevant for custom implementations. How strongly do you feel about it. It's public API so if there's a consensus that this needs a different name I'll suck it up 😄